In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import f_oneway
import warnings
warnings.filterwarnings('ignore')

In [2]:
aksidente = pd.read_csv('dataset\\accident_data.csv')
aksidente

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26/08/2019  51.517715   
3       200701BS66128           Serious    16/08/2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18/02/2022  57.374005   
660675  201091NM01881            Slight    21/02/2022  57.232273   
660676  201091NM01935            Slight    23/02/2022  57.585044   
660677  201091NM01964           Serious    23/02/2022  57.214898   
660678  201091NM02142           Serious    28/02/2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [3]:
aksidente['Accident Date'] = pd.to_datetime(aksidente['Accident Date'], dayfirst = True, errors = 'coerce')

In [4]:
aksidente['Accident Date']

0        2019-06-05
1        2019-07-02
2        2019-08-26
3        2019-08-16
4        2019-09-03
            ...    
660674   2022-02-18
660675   2022-02-21
660676   2022-02-23
660677   2022-02-23
660678   2022-02-28
Name: Accident Date, Length: 660679, dtype: datetime64[ns]

<h1>FILLING OUT NULL VALUES</h1>

In [5]:
aksidente.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

<h1>DATA CLEANSING</h1>

In [6]:
aksidente['Latitude'] = aksidente['Latitude'].fillna(aksidente['Latitude'].mode()[0])
aksidente['Longitude'] = aksidente['Longitude'].fillna(aksidente['Longitude'].mode()[0])
aksidente['Road_Surface_Conditions'] = aksidente['Road_Surface_Conditions'].fillna('unknown surface conditions')
aksidente['Road_Type'] = aksidente['Road_Type'].fillna('unaccounted')
aksidente['Urban_or_Rural_Area'] = aksidente['Urban_or_Rural_Area'].fillna(aksidente['Urban_or_Rural_Area'].mode()[0])
aksidente['Weather_Conditions'] = aksidente['Weather_Conditions'].fillna('unaccounted')

In [7]:
aksidente['Latitude'] = aksidente['Latitude'].astype('category')
aksidente['Longitude'] = aksidente['Longitude'].astype('category')
aksidente['Index'] = aksidente['Index'].astype('category')
aksidente['Accident_Severity'] = aksidente['Accident_Severity'].astype('category')
aksidente['Light_Conditions'] = aksidente['Light_Conditions'].astype('category')
aksidente['District Area'] = aksidente['District Area'].astype('category')
aksidente['Road_Surface_Conditions'] = aksidente['Road_Surface_Conditions'].astype('category')
aksidente['Road_Type'] = aksidente['Road_Type'].astype('category')
aksidente['Urban_or_Rural_Area'] = aksidente['Urban_or_Rural_Area'].astype('category')
aksidente['Weather_Conditions'] = aksidente['Weather_Conditions'].astype('category')
aksidente['Vehicle_Type'] = aksidente['Vehicle_Type'].astype('category')

In [8]:
aksidente.dtypes

Index                            category
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                         category
Light_Conditions                 category
District Area                    category
Longitude                        category
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
dtype: object

In [9]:
aksidente.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

<h1>Extracting date information using pandas date time</h1>

In [10]:
aksidente['Year'] = aksidente['Accident Date'].dt.year
aksidente['Month'] = aksidente['Accident Date'].dt.month
aksidente['Day'] = aksidente['Accident Date'].dt.day
aksidente['DayOfWeek'] = aksidente['Accident Date'].dt.dayofweek
aksidente

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    2019-06-05  51.506187   
1       200701BS65737           Serious    2019-07-02  51.495029   
2       200701BS66127           Serious    2019-08-26  51.517715   
3       200701BS66128           Serious    2019-08-16  51.495478   
4       200701BS66837            Slight    2019-09-03  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    2022-02-18  57.374005   
660675  201091NM01881            Slight    2022-02-21  57.232273   
660676  201091NM01935            Slight    2022-02-23  57.585044   
660677  201091NM01964           Serious    2022-02-23  57.214898   
660678  201091NM02142           Serious    2022-02-28  57.575210   

                   Light_Conditions           District Area Longitude  \
0             Darkness - lights lit  Kensington and Chelsea -0.209082   
1                          Daylight  Kensington and Chelsea -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea -0.210215   
3                          Daylight  Kensington and Chelsea -0.202731   
4             Darkness - lights lit  Kensington and Chelsea -0.192487   
...                             ...                     ...       ...   
660674                     Daylight                Highland -3.467828   
660675       Darkness - no lighting                Highland -3.809281   
660676                     Daylight                Highland -3.862727   
660677       Darkness - no lighting                Highland -3.823997   
660678                     Daylight                Highland -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2              unaccounted               Urban            unaccounted   
3       Single carriageway               Urban     Fine no high winds   
4              unaccounted               Urban            unaccounted   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  Year  Month  Day  DayOfWeek  
0                                        Car  2019      6    5          2  
1                                        Car  2019      7    2          1  
2                      Taxi/Private hire car  2019      8   26          0  
3       Bus or coach (17 or more pass seats)  2019      8   16          4  
4                              Other vehicle  2019      9    3          1  
...                                      ...   ...    ...  ...        ...  


<h1>------------------------------------------------------------</h1><h1 style="background-color: yellow; font-weight:bold;">Additional insights with Continuity<h6>(Aside from the other 20)</h6></h1>

<h4>Data Frame of the accidents in the month of January in the year 2019</h4>

In [11]:
df_2019_1 = aksidente[(aksidente['Year'] == 2019) & (aksidente['Month'] == 1)]
df_2019_1

Index Accident_Severity Accident Date   Latitude  \
22      200701BS70001            Slight    2019-01-18  51.499983   
23      200701BS70002           Serious    2019-01-22  51.485674   
24      200701BS70003           Serious    2019-01-17  51.498157   
25      200701BS70004            Slight    2019-01-03  51.493311   
26      200701BS70005            Slight    2019-01-02  51.498060   
...               ...               ...           ...        ...   
181974       2.01E+12            Slight    2019-01-23  55.197742   
181975       2.01E+12           Serious    2019-01-23  55.332789   
181976       2.01E+12            Slight    2019-01-28  54.984731   
181977       2.01E+12           Serious    2019-01-28  54.994760   
181978       2.01E+12           Serious    2019-01-31  55.196230   

              Light_Conditions           District Area Longitude  \
22       Darkness - lights lit  Kensington and Chelsea -0.179933   
23                    Daylight  Kensington and Chelsea -0.150543   
24                    Daylight  Kensington and Chelsea -0.201473   
25       Darkness - lights lit  Kensington and Chelsea -0.190427   
26                    Daylight  Kensington and Chelsea -0.166178   
...                        ...                     ...       ...   
181974                Daylight   Dumfries and Galloway -3.410764   
181975                Daylight   Dumfries and Galloway -3.443778   
181976  Darkness - no lighting   Dumfries and Galloway -3.166207   
181977  Darkness - no lighting   Dumfries and Galloway -3.224804   
181978                Daylight   Dumfries and Galloway -3.329328   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
22                         1                   1                     Dry   
23                         2                   2                     Dry   
24                         1                   1             Wet or damp   
25                         2                   1             Wet or damp   
26                         1                   1                     Dry   
...                      ...                 ...                     ...   
181974                     1                   2                     Dry   
181975                     1                   1             Wet or damp   
181976                     1                   1                     Dry   
181977                     1                   1                     Dry   
181978                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
22      Single carriageway               Urban     Fine no high winds   
23      Single carriageway               Urban     Fine no high winds   
24        Dual carriageway               Urban  Raining no high winds   
25      Single carriageway               Urban                  Other   
26      Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
181974    Dual carriageway               Rural     Fine no high winds   
181975  Single carriageway               Rural  Snowing no high winds   
181976  Single carriageway               Rural     Fine no high winds   
181977  Single carriageway               Rural     Fine no high winds   
181978  Single carriageway               Rural     Fine no high winds   

                                 Vehicle_Type  Year  Month  Day  DayOfWeek  
22                                        Car  2019      1   18          4  
23                                        Car  2019      1   22          1  
24                                        Car  2019      1   17          3  
25                                        Car  2019      1    3          3  
26                      Taxi/Private hire car  2019      1    2          2  
...                                       ...   ...    ...  ...        ...  
181974    Van / Goods 3.5 tonnes mgw or under  2019  

In [12]:
# accident severity among different areas
fatality= aksidente.groupby(['District Area','Accident_Severity'])['Number_of_Casualties'].size()
fatality.unstack()

Accident_Severity  Fatal  Serious  Slight
District Area                            
Aberdeen City         12      239    1072
Aberdeenshire         66      463    1401
Adur                   8      101     510
Allerdale             24      143     961
Alnwick                6       33     193
...                  ...      ...     ...
Wychavon              30      193    1138
Wycombe               20      216    1493
Wyre                  15      186    1037
Wyre Forest           22      132     815
York                  21      255    1621

[422 rows x 3 columns]

In [13]:
f_accident= aksidente.groupby(['Accident Date','Accident_Severity'])['Number_of_Casualties'].count()
f_accident

Accident Date  Accident_Severity
2019-01-01     Fatal                 11
               Serious               53
               Slight               271
2019-01-02     Fatal                  6
               Serious               57
                                   ... 
2022-12-30     Serious               21
               Slight               150
2022-12-31     Fatal                  4
               Serious               24
               Slight               165
Name: Number_of_Casualties, Length: 4383, dtype: int64

In [14]:
aksidente.describe().T
aksidente.groupby(['Accident_Severity','Vehicle_Type'])['Number_of_Casualties'].describe().T.head()

Accident_Severity                Fatal                                       \
Vehicle_Type      Agricultural vehicle Bus or coach (17 or more pass seats)   
count                        21.000000                           325.000000   
mean                          1.761905                             2.101538   
std                           1.044259                             3.618219   
min                           1.000000                             1.000000   
25%                           1.000000                             1.000000   

Accident_Severity                                             \
Vehicle_Type               Car Goods 7.5 tonnes mgw and over   
count              6577.000000                     216.00000   
mean                  1.891744                       1.87500   
std                   1.776131                       1.74892   
min                   1.000000                       1.00000   
25%                   1.000000                       1.00000   

Accident_Severity                                  \
Vehicle_Type      Goods over 3.5t. and under 7.5t   
count                                   67.000000   
mean                                     2.164179   
std                                      2.711351   
min                                      1.000000   
25%                                      1.000000   

Accident_Severity                                                              \
Vehicle_Type      Minibus (8 - 16 passenger seats) Motorcycle 125cc and under   
count                                    29.000000                 189.000000   
mean                                      2.517241                   1.767196   
std                                       2.339424                   1.166365   
min                                       1.000000                   1.000000   
25%                                       1.000000                   1.000000   

Accident_Severity                            \
Vehicle_Type      Motorcycle 50cc and under   
count                             95.000000   
mean                               1.726316   
std                                1.170943   
min                                1.000000   
25%                                1.000000   

Accident_Severity                                                              \
Vehicle_Type      Motorcycle over 125cc and up to 500cc Motorcycle over 500cc   
count                                        105.000000            339.000000   
mean                                           1.809524              1.793510   
std                                            1.233314              1.379848   
min                                            1.000000              1.000000   
25%                                            1.000000              1.000000   

Accident_Severity  ...                           Slight  \
Vehicle_Type       ... Minibus (8 - 16 passenger seats)   
count              ...                      1671.000000   
mean               ...                         1.323160   
std                ...                         0.724876   
min                ...                         1.000000   
25%                ...                         1.000000   

Accident_Severity                                                       \
Vehicle_Type      Motorcycle 125cc and under Motorcycle 50cc and under   
count                           13049.000000               6494.000000   
mean                                1.308836                  1.320450   
std                                 0.728955                  0.714651   
min                                 1.000000                  1.000000   
25%                                 1.000000                  1.000000   

Accident_Severity                                                              \
Vehicle_Type      Motorcycle over 125cc and up to 500cc Motorcycle over 500cc   
count                                       6537.000000         

In [15]:
aksidente.groupby(['Accident_Severity','Vehicle_Type'])['Number_of_Casualties'].agg(['count', 'mean', 'max'])

count      mean  \
Accident_Severity Vehicle_Type                                              
Fatal             Agricultural vehicle                       21  1.761905   
                  Bus or coach (17 or more pass seats)      325  2.101538   
                  Car                                      6577  1.891744   
                  Data missing or out of range                0       NaN   
                  Goods 7.5 tonnes mgw and over             216  1.875000   
                  Goods over 3.5t. and under 7.5t            67  2.164179   
                  Minibus (8 - 16 passenger seats)           29  2.517241   
                  Motorcycle 125cc and under                189  1.767196   
                  Motorcycle 50cc and under                  95  1.726316   
                  Motorcycle over 125cc and up to 500cc     105  1.809524   
                  Motorcycle over 500cc                     339  1.793510   
                  Other vehicle                              70  1.657143   
                  Pedal cycle                                 6  1.666667   
                  Ridden horse                                0       NaN   
                  Taxi/Private hire car                     155  1.987097   
                  Van / Goods 3.5 tonnes mgw or under       467  2.072805   
Serious           Agricultural vehicle                      282  1.475177   
                  Bus or coach (17 or more pass seats)     3373  1.471687   
                  Car                                     66461  1.469855   
                  Data missing or out of range                0       NaN   
                  Goods 7.5 tonnes mgw and over            2321  1.443774   
                  Goods over 3.5t. and under 7.5t           857  1.526254   
                  Minibus (8 - 16 passenger seats)          276  1.358696   
                  Motorcycle 125cc and under               2031  1.445101   
                  Motorcycle 50cc and under                1014  1.408284   
                  Motorcycle over 125cc and up to 500cc    1014  1.434911   
                  Motorcycle over 500cc                    3457  1.449523   
                  Other vehicle                             767  1.487614   
                  Pedal cycle                                39  1.256410   
                  Ridden horse                                0       NaN   
                  Taxi/Private hire car                    1771  1.488989   
                  Van / Goods 3.5 tonnes mgw or under      4554  1.467062   
Slight            Agricultural vehicle                     1644  1.313869   
                  Bus or coach (17 or more pass seats)    22180  1.319567   
                  Car                                    424954  1.333231   
                  Data missing or out of range                6  1.500000   
                  Goods 7.5 tonnes mgw and over           14770  1.329790   
                  Goods over 3.5t. and under 7.5t          5172  1.325406   
                  Minibus (8 - 16 passenger seats)         1671  1.323160   
                  Motorcycle 125cc and under              13049  1.308836   
                  Motorcycle 50cc and under                6494  1.320450   
                  Motorcycle over 125cc and up to 500cc    6537  1.321860   
                  Motorcycle over 500cc                   21861  1.338457   
                  Other vehicle                            4800  1.311875   
                  Pedal cycle                               152  1.388158   
                  Ridden horse                                4  1.250000   
                  Taxi/Private hire car                   11368  1.341485   
                  Van / Goods 3.5 tonnes mgw or under     29139  1.325440   

                                                          max  
Accident_Severity Vehicle_Type                                 
Fatal             Agricultural vehicle                    4.0  
               

<h1 style="font-weight:bold;">Univariate Analysis</h1>
<h1>What could be the light condition that has the most number of accidents in the first month of 2019?</h1>
<h4>majority of the accident happens on broad daylight. About 54.75% of the total accidents on January 2019 happened during Daylight</h4>

In [16]:
print(f"The total number of accident in the first month of 2019 is {df_2019_1 ['Light_Conditions'].value_counts().sum()}")
df_2019_1 ['Light_Conditions'].value_counts()

The total number of accident in the first month of 2019 is 15355


Light_Conditions
Daylight                       8407
Darkness - lights lit          5031
Darkness - no lighting         1585
Darkness - lighting unknown     248
Darkness - lights unlit          84
Name: count, dtype: int64

In [17]:
# Percentage Calculator Formula
v = ( df_2019_1 ['Light_Conditions'].value_counts().max() / df_2019_1 ['Light_Conditions'].value_counts().sum() ) * 100
np.round(v,2)

54.75

<h1>Does Year and Number of Casualties have correlation with one another?</h1><h4>Year does not have correlation with any number of casualties</h4>

In [18]:
correl = aksidente['Year'].corr(aksidente['Number_of_Casualties']) > 0.5
print(f"Year does not have any correlation with the number of casualties ({aksidente['Year'].corr(aksidente['Number_of_Casualties'])})")
correl

Year does not have any correlation with the number of casualties (-0.001549810694958303)


False

<h1 style="font-weight:bold;">Bivariate Analysis with Continuity</h1><h4>Data Frame of the accidents in the district area of Kensington and Chelsea in Daylight conditions</h4>

In [19]:
df_ken_day = aksidente[(aksidente['Light_Conditions'] == 'Daylight') & (aksidente['District Area'] == 'Kensington and Chelsea')]
df_ken_day

Index Accident_Severity Accident Date   Latitude  \
1       200701BS65737           Serious    2019-07-02  51.495029   
3       200701BS66128           Serious    2019-08-16  51.495478   
5       200701BS67159           Serious    2019-09-18  51.497750   
6       200701BS67207           Serious    2019-09-05  51.501405   
9       200701BS67543            Slight    2019-10-18  51.484539   
...               ...               ...           ...        ...   
532829  201001TB00417           Serious    2022-07-23  51.491188   
532927  201001TB00557            Slight    2022-08-01  51.502044   
533039  201001TB00720           Serious    2022-10-15  51.526104   
533316  201001TE00005           Serious    2022-01-17  51.520998   
533327  201001TE00016           Serious    2022-02-02  51.497193   

       Light_Conditions           District Area Longitude  \
1              Daylight  Kensington and Chelsea -0.173647   
3              Daylight  Kensington and Chelsea -0.202731   
5              Daylight  Kensington and Chelsea -0.192556   
6              Daylight  Kensington and Chelsea -0.161577   
9              Daylight  Kensington and Chelsea -0.175651   
...                 ...                     ...       ...   
532829         Daylight  Kensington and Chelsea -0.186910   
532927         Daylight  Kensington and Chelsea -0.191090   
533039         Daylight  Kensington and Chelsea -0.211905   
533316         Daylight  Kensington and Chelsea -0.213258   
533327         Daylight  Kensington and Chelsea -0.168230   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
1                          1                   2             Wet or damp   
3                          1                   4                     Dry   
5                          2                   3                     Dry   
6                          1                   2                     Dry   
9                          1                   2                     Dry   
...                      ...                 ...                     ...   
532829                     1                   2                     Dry   
532927                     1                   1             Wet or damp   
533039                     1                   1                     Dry   
533316                     1                   1                     Dry   
533327                     1                   2             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
1       Single carriageway               Urban  Raining no high winds   
3       Single carriageway               Urban     Fine no high winds   
5       Single carriageway               Urban     Fine no high winds   
6         Dual carriageway               Urban     Fine no high winds   
9       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
532829  Single carriageway               Urban     Fine no high winds   
532927  Single carriageway               Urban  Raining no high winds   
533039  Single carriageway               Urban     Fine no high winds   
533316  Single carriageway               Urban     Fine no high winds   
533327    Dual carriageway               Urban  Raining no high winds   

                                 Vehicle_Type  Year  Month  Day  DayOfWeek  
1                                         Car  2019      7    2          1  
3        Bus or coach (17 or more pass seats)  2019      8   16          4  
5                                         Car  2019      9   18          2  
6         Van / Goods 3.5 tonnes mgw or under  2019      9    5          3  
9       Motorcycle over 125cc and up to 500cc  2019     10   18          4  
...                                       ...   ...    ...  ...        ...  
532829                                    Car  2022      7   23          5  
532927                                    Car  2022      8  


<h1>How many casualties are involved in daylight conditions under the area of Keningston and Chelsea</h1>
<h4>There are 1989 casualties involved in daylight conditions under the area of Keningston and Chelsea</h4>

In [20]:
df_ken_day['Number_of_Casualties'].count()

1989

<h1> What type of accident severity is the most dominant in daylight conditions under the area of Keningston and Chelsea </h1><h4>The majority of accidents are slight 86.93% (1729), followed by serious 12.47% (248) and fatal 0.6% (12)</h4>

In [21]:
sum_of_all_accidents = (df_ken_day[df_ken_day['Accident_Severity'] == 'Slight'].count().mean() + df_ken_day[df_ken_day['Accident_Severity'] == 'Serious'].count().mean() + df_ken_day[df_ken_day['Accident_Severity'] == 'Fatal'].count().mean())
percentage_slight = np.round(((df_ken_day[df_ken_day['Accident_Severity'] == 'Slight'].count().mean() / sum_of_all_accidents) * 100),2)
percentage_serious = np.round(((df_ken_day[df_ken_day['Accident_Severity'] == 'Serious'].count().mean() / sum_of_all_accidents) * 100),2)
percentage_fatal = np.round(((df_ken_day[df_ken_day['Accident_Severity'] == 'Fatal'].count().mean() / sum_of_all_accidents) * 100),2)
print(f"The majority of accidents are slight {percentage_slight}%, followed by serious {percentage_serious}%, and fatal {percentage_fatal}%")
df_ken_day['Accident_Severity'].value_counts()


The majority of accidents are slight 86.93%, followed by serious 12.47%, and fatal 0.6%


Accident_Severity
Slight     1729
Serious     248
Fatal        12
Name: count, dtype: int64

<h1> Is there a correlation between the number of vehicles and number of casualties in the area of Keningston and Chelsea under daylight conditions? </h1><h4>There is no correlation between the number of vehicles and number of casualties (0.09)</h4>

In [22]:
df_ken_day['Number_of_Vehicles'].corr(df_ken_day['Number_of_Casualties']) > 0.5
print(f"There is no correlation between the number of vehicles and number of casualties {np.round(df_ken_day['Number_of_Vehicles'].corr(df_ken_day['Number_of_Casualties']),2)}")

There is no correlation between the number of vehicles and number of casualties 0.09


<h1> Are casualties greater in Sundays than Saturdays?</h1><h4>It appears that casualties are greater in Sundays than Saturdays</h4>

In [23]:
saturday_accidents = aksidente[(aksidente['DayOfWeek'] == 5)]['Number_of_Casualties']
sunday_accidents = aksidente[(aksidente['DayOfWeek'] == 6)]['Number_of_Casualties']

saturday_accidents.mean() < sunday_accidents.mean()

True

<h1> Are accident severity greater in Urban areas rather than Rural areas?</h1><h4>It appears that accident severity are greater in Rural than Urban, whilst Fatal being the most dominant accident severity</h4>

In [24]:
w = aksidente.groupby(['Accident_Severity','Urban_or_Rural_Area'])['Number_of_Casualties'].mean()
np.round(w,2)

Accident_Severity  Urban_or_Rural_Area
Fatal              Rural                  2.09
                   Unallocated             NaN
                   Urban                  1.55
Serious            Rural                  1.68
                   Unallocated            1.00
                   Urban                  1.31
Slight             Rural                  1.42
                   Unallocated            1.20
                   Urban                  1.28
Name: Number_of_Casualties, dtype: float64

<h4>It appears that accident severity are greater in Rural than Urban</h4>

<h1 style="font-weight:bold;">Multivariate Analysis with Continuity</h1><h1>In daylight conditions according to the data, how many are the accidents, its average, and the maximum number of accident in various roadtype?</h1>

<h4>Dual carriageways have the highest average number of accidents (1.462943).</h4>

In [25]:
daylight_accidents = aksidente[(aksidente['Light_Conditions'] == 'Daylight')]
daylight_accidents

Index Accident_Severity Accident Date   Latitude  \
1       200701BS65737           Serious    2019-07-02  51.495029   
3       200701BS66128           Serious    2019-08-16  51.495478   
5       200701BS67159           Serious    2019-09-18  51.497750   
6       200701BS67207           Serious    2019-09-05  51.501405   
9       200701BS67543            Slight    2019-10-18  51.484539   
...               ...               ...           ...        ...   
660672  201091NM01085           Serious    2022-01-29  57.170725   
660673  201091NM01190            Slight    2022-02-02  57.585152   
660674  201091NM01760            Slight    2022-02-18  57.374005   
660676  201091NM01935            Slight    2022-02-23  57.585044   
660678  201091NM02142           Serious    2022-02-28  57.575210   

       Light_Conditions           District Area Longitude  \
1              Daylight  Kensington and Chelsea -0.173647   
3              Daylight  Kensington and Chelsea -0.202731   
5              Daylight  Kensington and Chelsea -0.192556   
6              Daylight  Kensington and Chelsea -0.161577   
9              Daylight  Kensington and Chelsea -0.175651   
...                 ...                     ...       ...   
660672         Daylight                Highland -3.852421   
660673         Daylight                Highland -3.748312   
660674         Daylight                Highland -3.467828   
660676         Daylight                Highland -3.862727   
660678         Daylight                Highland -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
1                          1                   2             Wet or damp   
3                          1                   4                     Dry   
5                          2                   3                     Dry   
6                          1                   2                     Dry   
9                          1                   2                     Dry   
...                      ...                 ...                     ...   
660672                     1                   2                    Snow   
660673                     2                   2             Wet or damp   
660674                     2                   1                     Dry   
660676                     1                   3            Frost or ice   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
1       Single carriageway               Urban  Raining no high winds   
3       Single carriageway               Urban     Fine no high winds   
5       Single carriageway               Urban     Fine no high winds   
6         Dual carriageway               Urban     Fine no high winds   
9       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660672  Single carriageway               Rural     Fine no high winds   
660673  Single carriageway               Rural     Fine no high winds   
660674  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                 Vehicle_Type  Year  Month  Day  DayOfWeek  
1                                         Car  2019      7    2          1  
3        Bus or coach (17 or more pass seats)  2019      8   16          4  
5                                         Car  2019      9   18          2  
6         Van / Goods 3.5 tonnes mgw or under  2019      9    5          3  
9       Motorcycle over 125cc and up to 500cc  2019     10   18          4  
...                                       ...   ...    ...  ...        ...  
660672                                    Car  2022      1   29          5  
660673                                    Car  2022      2  

In [26]:
count_dual = len(daylight_accidents[daylight_accidents['Road_Type'] == 'Dual carriageway'])
sum_casualties_dual = daylight_accidents[daylight_accidents['Road_Type'] == 'Dual carriageway']['Number_of_Casualties'].sum()
sum_casualties_dual / count_dual

1.4629434841937685

<h4>"Single carriageway" has the highest number of accidents by far, with 363,265.</h4>

In [27]:
daylight_accidents[daylight_accidents['Road_Type'] == 'Single carriageway']['Number_of_Casualties'].count()

363265

<h4>Single carriageways also have the highest maximum number of accidents.</h4>

In [28]:
daylight_accidents[daylight_accidents['Road_Type'] == 'Single carriageway']['Number_of_Casualties'].max()

62

In [29]:
daylight_accidents = aksidente[aksidente['Light_Conditions'] == 'Daylight']
road_type_casualties = daylight_accidents.groupby('Road_Type')['Number_of_Casualties'].agg(['count', 'mean', 'max'])
road_type_casualties

count      mean  max
Road_Type                                
Dual carriageway     70352  1.462943   47
One way street        9964  1.177439   15
Roundabout           32581  1.257604   19
Single carriageway  363265  1.328810   62
Slip road             5249  1.394551   14
unaccounted           3469  1.219948   28

<h1>------------------------------------------------------------</h1><h1 style="font-weight:bold; background-color: yellow;"> 20 insights </h1>

<h3>#1 Univariate Analysis</h3><h1>WHAT ARE THE TOTAL NUMBER OF ACCIDENTS</h1><h4>There are 660679 total number of accidents</h4>

In [30]:
a = aksidente['Index'].count()
a

660679

<h3>#2 Univariate Analysis</h3><h1>AVERAGE NUMBER OF CASUALTIES PER ACCIDENT</h1><h4>The mean number of casualities is 1.36</h4>

In [31]:
b = aksidente['Number_of_Casualties'].mean()
np.round(b,2)

1.36

<h3>#3 Univariate Analysis</h3><h1>MOST COMMON ACCIDENT SEVERITY</h1><h4>The most comon accident severity is 'Slight'</h4>

In [32]:
c =aksidente['Accident_Severity'].mode()[0]
c

'Slight'

<h3>#4 Univariate Analysis</h3><h1>DISTRIBUTION OF ROAD TYPES</h1><h4>The majority of accidents occur on single carriageways (492,143), followed by dual carriageways (99,424), roundabouts (43,992), one-way streets (13,559), slip roads (7,041), and unaccounted (4,520)</h4>

In [33]:
d = aksidente['Road_Type'].value_counts()
d

Road_Type
Single carriageway    492143
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
unaccounted             4520
Name: count, dtype: int64

<h3>#5 Univariate Analysis</h3><h1>MOST FREQUENT WEATHER CONDITION DURING ACCIDENTS</h1><h4>The most frequent weather condition during accidents is 'Fine no high winds'</h4>

In [34]:
e = aksidente['Weather_Conditions'].mode()[0]
e

'Fine no high winds'

<h3>#6 Univariate Analysis</h3><h1>AVERAGE NUMBER OF VEHICLES INVOLVED IN ACCIDENTS</h1><h4>The average number of vehicles involved in accidents is 1.83</h4>

In [35]:
f = aksidente['Number_of_Vehicles'].mean()
np.round(f,2)

1.83

<h3>#7 Univariate Analysis</h3><h1>PERCENTAGE OF ACCIDENTS IN URBAN VS. RURAL AREAS</h1><h4>Approximately 63.83% of accidents occur in urban areas, 36.17% in rural areas, and a negligible 0.002% in unallocated areas.</h4>

In [36]:
g = aksidente['Urban_or_Rural_Area'].value_counts(normalize=True) * 100
np.round(g,3)

Urban_or_Rural_Area
Urban          63.825
Rural          36.173
Unallocated     0.002
Name: proportion, dtype: float64

<h3>#8 Univariate Analysis</h3><h1>MOST COMMON LIGHT CONDITION DURING ACCIDENTS</h1><h4>The most common light condition during accidents is 'Daylight'</h4>

In [37]:
h = aksidente['Light_Conditions'].mode()[0]
h

'Daylight'

<h3>#9 Univariate Analysis</h3><h1>DISTRIBUTION OF ROAD SURFACE CONDITIONS</h1><h4>Most accidents occur on dry road surfaces (447,821), followed by wet or damp surfaces (186,708), frost or ice (18,517), snow (5,890), flood over 3cm deep (1,017), and unknown surface conditions (726)</h4>

In [38]:
i = aksidente['Road_Surface_Conditions'].value_counts()
i

Road_Surface_Conditions
Dry                           447821
Wet or damp                   186708
Frost or ice                   18517
Snow                            5890
Flood over 3cm. deep            1017
unknown surface conditions       726
Name: count, dtype: int64

<h3>#10 Univariate Analysis</h3><h1>MOST COMMON VEHICLE TYPE INVOLVED IN ACCIDENTS</h1><h4>The most common vehicle type involved in accidents is 'Car'</h4>

In [39]:
j = aksidente['Vehicle_Type'].mode()[0]
j

'Car'

<h3>#11  Univariate Analysis</h3><h1>AVERAGE NUMBER OF CASUALTIES BY ACCIDENT SEVERITY</h1><h4>Fatal accidents have the highest average number of casualties (1.90), followed by serious accidents (1.47) and slight accidents (1.33).</h4>

In [40]:
k = aksidente.groupby('Accident_Severity')['Number_of_Casualties'].mean()
np.round(k,2)

Accident_Severity
Fatal      1.90
Serious    1.47
Slight     1.33
Name: Number_of_Casualties, dtype: float64

<h3>#12 Univariate Analysis</h3><h1>TOP 5 NUMBER OF ACCIDENTS BY DISTRICT AREA</h1><h4>Birmingham has the highest number of accidents (13,491), followed by Leeds (8,898), Manchester (6,720), Bradford (6,212), and Sheffield (5,710)</h4>

In [41]:
l = aksidente['District Area'].value_counts().head()
l

District Area
Birmingham    13491
Leeds          8898
Manchester     6720
Bradford       6212
Sheffield      5710
Name: count, dtype: int64

<h3>#13 Univariate Analysis</h3><h1>CORRELATION BETWEEN NUMBER OF VEHICLES AND NUMBER OF CASUALTIES</h1><h4>There is no correlation between the number of vehicles and the number of casualties</h4>

In [42]:
m = aksidente['Number_of_Vehicles'].corr(aksidente['Number_of_Casualties']) > 0.5
m

False

<h3>#14 Univariate Analysis</h3><h1>AVERAGE NUMBER OF CASUALTIES IN URBAN VS. RURAL AREAS</h1><h4>Rural areas have a slightly higher average number of casualties per accident (1.48) compared to urban areas (1.29). Unallocated areas have an average of 1.18 casualties per accident.</h4>

In [43]:
n = aksidente.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].mean()
np.round(n,2)

Urban_or_Rural_Area
Rural          1.48
Unallocated    1.18
Urban          1.29
Name: Number_of_Casualties, dtype: float64

<h3>#15 Univariate Analysis</h3><h1>MOST COMMON ROAD TYPE FOR EACH ACCIDENT SEVERITY</h1><h4>Single carriageways are the most common road type for all accident severities: fatal (6,527), serious (70,059), and slight (415,557)</h4>

In [44]:
o = aksidente.groupby(['Accident_Severity', 'Road_Type']).size().unstack()
o

Road_Type          Dual carriageway  One way street  Roundabout  \
Accident_Severity                                                 
Fatal                          1815              95         142   
Serious                       11746            1655        3665   
Slight                        85863           11809       40185   

Road_Type          Single carriageway  Slip road  unaccounted  
Accident_Severity                                              
Fatal                            6527         49           33  
Serious                         70059        611          481  
Slight                         415557       6381         4006

<h3>#16 Univariate Analysis</h3><h1>NUMBER OF ACCIDENTS BY WEATHER CONDITIONS</h1><h4>The majority of accidents occur during "Fine no high winds" (520,885), followed by "Raining no high winds" (79,696), and "Other" conditions (17,150)</h4>

In [45]:
p = aksidente['Weather_Conditions'].value_counts()
p

Weather_Conditions
Fine no high winds       520885
Raining no high winds     79696
Other                     17150
unaccounted               14128
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h3>#17 Bivariate Analysis</h3><h1>AVERAGE NUMBER OF VEHICLES INVOLVED IN ACCIDENTS BY ROAD SURFACE CONDITION</h1><h4>Accidents on dry roads involve the highest average number of vehicles (1.84), followed by wet or damp roads roads (1.83), snow (1.75), unknown surface conditions (1.74), frost or ice (1.64), and flood over 3cm deep (1.61).</h4>

In [46]:
q = aksidente.groupby('Road_Surface_Conditions')['Number_of_Vehicles'].mean()
np.round(q,2)

Road_Surface_Conditions
Dry                           1.84
Flood over 3cm. deep          1.61
Frost or ice                  1.64
Snow                          1.75
Wet or damp                   1.83
unknown surface conditions    1.74
Name: Number_of_Vehicles, dtype: float64

<h3>#18 Bivariate Analysis</h3><h1>RANKING OF ACCIDENT SEVERITY DISTRIBUTION BY LIGHT CONDITIONS</h1><h4>Accidents in daylight have the highest severity (rank 1.0), darkness with lights lit (rank 2.0), darkness with no lighting (rank 3.0), and darkness with unknown lighting (rank 4.0), while those in darkness with unlit lights have the lowest severity (rank 5.0)</h4>

In [47]:
r = aksidente.groupby(['Light_Conditions', 'Accident_Severity']).size().unstack()
r.rank(ascending=False)

Accident_Severity            Fatal  Serious  Slight
Light_Conditions                                   
Darkness - lighting unknown    4.0      4.0     4.0
Darkness - lights lit          2.0      2.0     2.0
Darkness - lights unlit        5.0      5.0     5.0
Darkness - no lighting         3.0      3.0     3.0
Daylight                       1.0      1.0     1.0

<h3>#19 Multivariate Analysis</h3><h1>AVERAGE NUMBER OF CASUALTIES BY ROAD TYPE AND WEATHER CONDITIONS</h1><h4>The average number of casualties by road type and weather conditions indicates that dual carriageways have the highest average casualties in fog or mist (1.67), while one-way streets have the lowest average casualties in snowing with high winds (1.07). Other notable averages with high records include roundabouts in snowing with high winds (1.53), single carriageways in snow with high winds (1.43), and slip roads in raining with no high winds (1.48)</h4>

In [48]:
s = aksidente.groupby(['Road_Type', 'Weather_Conditions'])['Number_of_Casualties'].mean().unstack()
np.round(s,2)

Weather_Conditions  Fine + high winds  Fine no high winds  Fog or mist  Other  \
Road_Type                                                                       
Dual carriageway                 1.44                1.48         1.67   1.43   
One way street                   1.18                1.19         1.24   1.23   
Roundabout                       1.33                1.27         1.32   1.27   
Single carriageway               1.38                1.34         1.41   1.35   
Slip road                        1.36                1.42         1.39   1.46   
unaccounted                      1.18                1.25         1.52   1.17   

Weather_Conditions  Raining + high winds  Raining no high winds  \
Road_Type                                                         
Dual carriageway                    1.48                   1.50   
One way street                      1.31                   1.24   
Roundabout                          1.25                   1.29   
Single carriageway                  1.42                   1.40   
Slip road                           1.43                   1.48   
unaccounted                         1.32                   1.31   

Weather_Conditions  Snowing + high winds  Snowing no high winds  unaccounted  
Road_Type                                                                     
Dual carriageway                    1.39                   1.40         1.33  
One way street                      1.07                   1.23         1.14  
Roundabout                          1.53                   1.32         1.23  
Single carriageway                  1.43                   1.33         1.22  
Slip road                           1.30                   1.23         1.29  
unaccounted                         2.00                   1.40         1.22

<h3>#20 Multivariate Analysis</h3><h1>NUMBER OF ACCIDENTS BY DISTRICT AREA AND ACCIDENT SEVERITY</h1><h4>The number of accidents by district area and accident severity shows that York has the highest number of slight accidents (1,621), while Alnwick has the lowest number of fatal accidents (6). Other notable figures include Aberdeen City with 12 fatal accidents, Aberdeenshire with 463 serious accidents, and Wycombe with 1,493 slight accidents</h4>

In [49]:
t = aksidente.groupby(['District Area', 'Accident_Severity'])['Number_of_Casualties'].size().unstack()
t

Accident_Severity  Fatal  Serious  Slight
District Area                            
Aberdeen City         12      239    1072
Aberdeenshire         66      463    1401
Adur                   8      101     510
Allerdale             24      143     961
Alnwick                6       33     193
...                  ...      ...     ...
Wychavon              30      193    1138
Wycombe               20      216    1493
Wyre                  15      186    1037
Wyre Forest           22      132     815
York                  21      255    1621

[422 rows x 3 columns]